ca/certificate-authority.go

```diff
@@ -305,86 +408,246 @@ func (ca *CertificateAuthorityImpl) IssueCertificate(csr x509.CertificateRequest
-		return emptyCert, err
+		ca.Log.Audit(fmt.Sprintf("PEM empty from Signer, rolling back: serial=[%s] err=[%v]", serialHex, err))
+		ca.SA.UpdateCertificateRequestStatus(requestID, core.StatusInvalid)
+		return
 	}
 
 	block, _ := pem.Decode(certPEM)
 	if block == nil || block.Type != "CERTIFICATE" {
 		err = core.InternalServerError("Invalid certificate value returned")
 		// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
-		ca.log.Audit(fmt.Sprintf("PEM decode error, aborting and rolling back issuance: pem=[%s] err=[%v]", certPEM, err))
-		return emptyCert, err
+		ca.Log.Audit(fmt.Sprintf("PEM decode error, aborting and rolling back issuance: pem=[%s] err=[%v]", certPEM, err))
+		ca.SA.UpdateCertificateRequestStatus(requestID, core.StatusInvalid)
+		return
 	}
 	certDER := block.Bytes
 
-	cert := core.Certificate{
-		DER: certDER,
-	}
-
 	// This is one last check for uncaught errors
 	if err != nil {
 		err = core.InternalServerError(err.Error())
 		// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
-		ca.log.Audit(fmt.Sprintf("Uncaught error, aborting and rolling back issuance: pem=[%s] err=[%v]", certPEM, err))
-		return emptyCert, err
+		ca.Log.Audit(fmt.Sprintf("Uncaught error, aborting and rolling back issuance: pem=[%s] err=[%v]", certPEM, err))
+		ca.SA.UpdateCertificateRequestStatus(requestID, core.StatusInvalid)
+		return
 	}
 
 	// Store the cert with the certificate authority, if provided
-	_, err = ca.SA.AddCertificate(certDER, regID)
+	cert, err := ca.SA.AddCertificate(certDER, requestID)
 	if err != nil {
 		err = core.InternalServerError(err.Error())
 		// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
-		ca.log.Audit(fmt.Sprintf("Failed RPC to store at SA, orphaning certificate: pem=[%s] err=[%v]", certPEM, err))
-		return emptyCert, err
+		ca.Log.Audit(fmt.Sprintf("Failed RPC to store at SA, orphaning certificate: pem=[%s] err=[%v]", certPEM, err))
+		ca.SA.UpdateCertificateRequestStatus(requestID, core.StatusInvalid)
+		return
 	}
 
-	// Submit the certificate to any configured CT logs
+	// Perform post-issuance checks
 	certObj, err := x509.ParseCertificate(certDER)
 	if err != nil {
-		ca.log.Warning(fmt.Sprintf("Post-Issuance OCSP failed parsing Certificate: %s", err))
-		return cert, nil
+		ca.Log.Warning(fmt.Sprintf("Post-Issuance OCSP failed parsing Certificate: %s", err))
+		return
+	}
+
+	err = ca.MatchesCSR(cert, csr)
+	if err != nil {
+		err = core.InternalServerError(err.Error())
+		// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
+		ca.Log.Audit(fmt.Sprintf("Failed RPC to store at SA, orphaning certificate: pem=[%s] err=[%v]", certPEM, err))
+		ca.SA.UpdateCertificateRequestStatus(requestID, core.StatusInvalid)
+		return
+	}
+
+	// Here's where we finally declare success
+	// Now that the certificate has been committed, we can fill out
+	// the rest of the log event and mark it as successful
+	logEvent.SerialNumber = cert.Serial
```

Comment:
I'd like this to be set when the serial's generated, so it gets emitted in the log event in exception cases